### PR TITLE
Escrow updates for ASv2

### PR DESCRIPTION
**Pull request by eelanagaraj**
## Description

- Implements changes for Escrow to accept payments requiring an attestation from a trustedIssuer
- See [this design doc](https://www.notion.so/clabsco/Escrow-Design-Doc-WIP-2cc7507125964ab48c5061bbba479010) for more details & discussion around proposed changes
- will require another update/check once lookup functions in `FederatedAttestations` are finalized

### Notes/Questions for Reviewers
- note: change in `registryContract.getAddressFor` -> `registryContract.getAddressForOrDie` in transfer (for attestations contract)
- is there a more sensible way to decide on/set governanceConstitution thresholds?

## Tested
- Unit tests

## Backwards Compatibility
Since this is based off of [this PR](https://github.com/celo-org/celo-monorepo/pull/9631) and moving from `UsingRegistry`-> `UsingRegistryV2`, this is a Major upgrade, since the `setRegistry` function is no longer part of the contract ABI. Other than that, the other function changes are backwards compatible.

### CODE REVIEW

File: packages/protocol/contracts/common/UsingRegistry.sol [hunk 1]
```diff
@@ -36,6 +37,9 @@ contract UsingRegistry is Ownable {
   bytes32 constant FEE_CURRENCY_WHITELIST_REGISTRY_ID = keccak256(
     abi.encodePacked("FeeCurrencyWhitelist")
   );
+  bytes32 constant FEDERATED_ATTESTATIONS_REGISTRY_ID = keccak256(
+    abi.encodePacked("FederatedAttestations")
```
Comment (isabellewei) [inline packages/protocol/contracts/common/UsingRegistry.sol:41]: why is this needed if the FederatedAttestations contract uses `UsingRegistryV2`?

File: packages/protocol/contracts/identity/Escrow.sol [hunk 2]
```diff
@@ -93,6 +106,46 @@ contract Escrow is
...
+        defaultTrustedIssuers[i] != trustedIssuer,
+        "trustedIssuer already in defaultTrustedIssuers"
+      );
+    }
+    defaultTrustedIssuers.push(trustedIssuer);
+    emit DefaultTrustedIssuerAdded(trustedIssuer);
+  }
+
+  /**
+   * @notice Add an address to the defaultTrustedIssuers list.
```
Comment (isabellewei) [inline packages/protocol/contracts/identity/Escrow.sol]: think you forgot to update this docstring :)

File: packages/protocol/contracts/identity/Escrow.sol [hunk 3]
```diff
@@ -103,9 +156,13 @@ contract Escrow is
...
   *        prove ownership of the corresponding private key to withdraw from escrow.
   * @param minAttestations The min number of attestations required to withdraw the payment.
+  * @return True if transfer succeeded.
   * @dev Throws if 'token' or 'value' is 0.
+  * @dev Throws if identifier is null and minAttestations > 0.
+  * @dev If minAttestations is 0, trustedIssuers will be set to empty list.
   * @dev msg.sender needs to have already approved this contract to transfer
-  * @dev If no identifier is given, then minAttestations must be 0.
+
+
```
Comment (isabellewei) [inline packages/protocol/contracts/identity/Escrow.sol:164]: nit: extra empty lines here

File: packages/protocol/contracts/identity/Escrow.sol [hunk 4]
```diff
@@ -93,6 +106,46 @@ contract Escrow is
...
+    if (index != numDefaultTrustedIssuers - 1) {
+      // Swap last index with index-to-remove
+      defaultTrustedIssuers[index] = defaultTrustedIssuers[numDefaultTrustedIssuers - 1];
+    }
+    defaultTrustedIssuers.pop();
+    emit DefaultTrustedIssuerRemoved(trustedIssuer);
+  }
+
   /**
   * @notice Transfer tokens to a specific user. Supports both identity with privacy (an empty
```
Comment (isabellewei) [inline packages/protocol/contracts/identity/Escrow.sol:149]: maybe add a note about how the the default trusted issuers list is used?

File: packages/protocol/contracts/identity/Escrow.sol [hunk 5]
```diff
@@ -249,6 +368,121 @@ contract Escrow is
...
+
+    EscrowedPayment storage newPayment = escrowedPayments[paymentId];
+    require(newPayment.timestamp == 0, "paymentId already used");
+    newPayment.recipientIdentifier = identifier;
+    newPayment.sender = msg.sender;
+    newPayment.token = token;
+    newPayment.value = value;
+    newPayment.sentIndex = sentIndex;
+    newPayment.receivedIndex = receivedIndex;
+    // solhint-disable-next-line not-rely-on-time
```
Comment (isabellewei) [inline packages/protocol/contracts/identity/Escrow.sol:504]: I think this is not needed if you used `block.timestamp` instead of `now`

File: packages/protocol/contracts/identity/Escrow.sol (hunk 2 above)
Comment (alecps) [inline packages/protocol/contracts/identity/Escrow.sol]: Nit: "Remove an address..."

File: packages/protocol/contracts/identity/Escrow.sol [hunk 6]
```diff
@@ -93,6 +106,46 @@ contract Escrow is
...
+    defaultTrustedIssuers.push(trustedIssuer);
+    emit DefaultTrustedIssuerAdded(trustedIssuer);
+  }
+
+  /**
+   * @notice Add an address to the defaultTrustedIssuers list.
+   * @param trustedIssuer Address of the trustedIssuer to remove.
+   * @param index Index of trustedIssuer in defaultTrustedIssuers.
+   * @dev Throws if trustedIssuer is not in defaultTrustedIssuers.
+   * @dev Throws if trustedIssuer is not in defaultTrustedIssuers.
```
Comment (alecps) [inline packages/protocol/contracts/identity/Escrow.sol]: duplicate lines

File: packages/protocol/contracts/identity/Escrow.sol [hunk 7]
```diff
@@ -168,17 +250,54 @@ contract Escrow is
     EscrowedPayment memory payment = escrowedPayments[paymentId];
     require(payment.token != address(0) && payment.value > 0, "Invalid withdraw value.");
 
+    bool passedCheck = false;
```
Comment (alecps) [inline packages/protocol/contracts/identity/Escrow.sol]: Nit: consider breaking this check into a separate view function that returns a boolean

Comment (alecps) [reply packages/protocol/contracts/identity/Escrow.sol]: I think this should allow you to get rid of the passedCheck boolean and all the break/continue/if-else's

File: packages/protocol/test/identity/escrow.ts [hunk 8]
```diff
@@ -96,16 +120,100 @@ contract('Escrow', (accounts: string[]) => {
...
+
+    it('should not allow a trustedIssuer to be added twice', async () => {
+      await escrow.addDefaultTrustedIssuer(trustedIssuer1, { from: owner })
+      await assertRevertWithReason(
+        escrow.addDefaultTrustedIssuer(trustedIssuer1, { from: owner }),
+        'trustedIssuer already in defaultTrustedIssuers'
+      )
+    })
+
+    it('should a second trustedIssuer to be added', async () => {
```
Comment (alecps) [inline packages/protocol/test/identity/escrow.ts]: should allow

File: packages/protocol/test/identity/escrow.ts [hunk 9]
```diff
@@ -96,16 +120,100 @@ contract('Escrow', (accounts: string[]) => {
...
+    it('should revert if trusted issuer does not match index', async () => {
+      await assertRevertWithReason(
+        escrow.removeDefaultTrustedIssuer(trustedIssuer2, 0, { from: owner }),
+        'trustedIssuer does not match address found at defaultTrustedIssuers[index]'
+      )
+    })
+
+    it('allows owner to remove trustedIssuer when two are present', async () => {
+      await escrow.addDefaultTrustedIssuer(trustedIssuer2, { from: owner })
+      assert.deepEqual(await escrow.getDefaultTrustedIssuers(), [trustedIssuer1, trustedIssuer2])
```
Comment (alecps) [inline packages/protocol/test/identity/escrow.ts:211]: Does deepEqual check ordering?

Comment (alecps) [reply packages/protocol/test/identity/escrow.ts:211]: Asking because we want to make sure we get the indexes right

**Review by alecps — APPROVED** [review]
Awesome work! The only thing I want to double check on is whether there are any anticipated headaches (e.g. with the escrow contract storage) if we decide to move the defaultTrustedIssuers list into a separate contract / token curated registry in the future

File: packages/protocol/test/identity/escrow.ts (hunk 9 above)
Comment (eelanagaraj) [reply packages/protocol/test/identity/escrow.ts:211]: Yes I believe it does! DeepEqual should basically compare that the object's properties/values are identical, but that they're not the same object (i.e. deepEqual([1,2,3], [1,2,3]) passes, deepEqual([1,2,3], [2,3,1]) fails, and (===) equal([1,2,3], [1,2,3]) fails

File: packages/protocol/contracts/identity/Escrow.sol (hunk 7 above)
Comment (eelanagaraj) [reply packages/protocol/contracts/identity/Escrow.sol]: I broke it into a couple of functions to hopefully help readability (will merge but please take a look in the big PR and lmk if it's more readable!); unfortunately I think we can't fully get rid of the passedCheck, in particular the if/else on trustedIssuers.length > 0, since we want v1Attestations to be checked only when specified as a trustedIssuer + for backwards compatibility (i.e. trustedIssuers not set), but not as a default/backup in case other attestations are not found. Maybe I'm misunderstanding you though, lmk if you meant something else 🙏

**Comment by eelanagaraj** [comment]
> Awesome work! The only thing I want to double check on is whether there are any anticipated headaches (e.g. with the escrow contract storage) if we decide to move the defaultTrustedIssuers list into a separate contract / token curated registry in the future

@alecps feel you! My hope is that since we are only using this list in `getDefaultTrustedIssuers` which is then called in `transfer` (which tbh as is may be overkill since it's so simple as is...), we can always update `getDefaultTrustedIssuers` to pull the list from elsewhere, including another contract, wdyt?

File: packages/protocol/contracts/identity/Escrow.sol [hunk 10]
```diff
@@ -249,6 +348,176 @@ contract Escrow is
     return sentPaymentIds[sender];
   }
 
+  /**
+  * @notice Gets array of all trusted issuers set per paymentId.
+  * @param paymentId The ID of the payment to be deleted.
+  * @return An array of addresses of trusted issuers set for an escrowed payment.
+  */
+  function getTrustedIssuersPerPayment(address paymentId) external view returns (address[] memory) {
+    return trustedIssuersPerPayment[paymentId];
```
Comment (isabellewei) [inline packages/protocol/contracts/identity/Escrow.sol:357]: couldn't we directly call this variable without creating this extra function?

File: packages/protocol/contracts/common/UsingRegistry.sol (hunk 1 above)
Comment (eelanagaraj) [reply packages/protocol/contracts/common/UsingRegistry.sol:41]: it's actually for other contracts that would want to interact with FederatedAttestations, but that said, they should use `UsingRegistryV2` in the future anyways 😅 will remove

File: packages/protocol/contracts/identity/Escrow.sol (hunk 10 above)
Comment (eelanagaraj) [reply packages/protocol/contracts/identity/Escrow.sol:357]: this just makes it easier to get the entire list when writing tests 😬 
bogdan mentioned this as well in the main PR though--will resolve here in favor of discussion there!

File: packages/protocol/contracts/identity/Escrow.sol (hunk 5 above)
Comment (eelanagaraj) [reply packages/protocol/contracts/identity/Escrow.sol:504]: this was a holdover from before but interestingly not needed in either case 😂 removing & changing to block.timestamp across the file anyways

File: packages/protocol/contracts/identity/Escrow.sol (hunk 2 above)
Comment (eelanagaraj) [reply packages/protocol/contracts/identity/Escrow.sol]: indeed! :D
